core/src/main/scala/chisel3/Data.scala

```diff
@@ -214,8 +214,28 @@ private[chisel3] object getRecursiveFields {
 private[chisel3] object getMatchedFields {
   def apply(x: Data, y: Data): Seq[(Data, Data)] = (x, y) match {
     case (x: Element, y: Element) =>
-      require(x.typeEquivalent(y))
+      require(
+        x.typeEquivalent(y), {
+          val reason = x
+            .findFirstTypeMismatch(y, strictTypes = true, strictWidths = true, strictProbeInfo = true)
+            .map(s => s"\nbecause: $s")
+            .getOrElse("")
+          s"$x is not typeEquivalent to $y $reason"
+        }
+      )
+      Seq(x -> y)
+    case (_, _) if DataMirror.hasProbeTypeModifier(x) || DataMirror.hasProbeTypeModifier(y) => {
+      require(
+        x.typeEquivalent(y), {
```

Comment:
These might motivate a `requireTypeEquivalent` or something? Fine as-is, just a thought.

Reply:
Awesome explanation messages!! 🎉 🚀

Reply:
yeah good point, i did find myself copying this everywhere

Reply:
added a method and used it in a few places, and just deleted this message in other places as it was more for debugging